bundle-workflow/src/s3_utility.py

```diff
@@ -0,0 +1,137 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import os
+from pathlib import Path
+from urllib.parse import urlparse
+
+import boto3
+from botocore.exceptions import ClientError
+
+
+class S3Bucket:
+    def __init__(self, bucket_name, role_arn=None, role_session_name=None):
+        """
+        Provides methods to download/upload files and folders to S3 bucket
+
+        :param bucket_name: The s3 bucket name
+        :param role_arn: the arn of the role that has permissions to access S3
+        :param role_session_name: the aws role session name
+        """
+        self.bucket_name = bucket_name
+        self.role_arn = (
+            role_arn
+            if role_arn is not None
+            else os.environ.get("AWS_ROLE_ARN")
+        )
+        self.role_session_name = (
+            role_session_name
+            if role_session_name is not None
+            else os.environ.get("AWS_ROLE_SESSION_NAME")
+        )
+        assumed_role_cred = self.__assume_role()
+        self.__s3_client = boto3.client(
```

Comment:
Nit: I would save `__ assumed_role_cred` and create `__s3_client` and `__s3_resource ` on demand the first time you need it.